packages/integrations/cloudflare/src/index.ts

```diff
@@ -90,33 +94,91 @@ export default function createIntegration(args?: Options): AstroIntegration {
 					vite.ssr.target = 'webworker';
 				}
 			},
+			// BUG: `entrypoints` currently does not respect SERVER_BUILD_FOLDER, aka `build.server`
+			'astro:build:ssr': ({ manifest, entryPoints }) => {
+				// NOTE: monkey-patch because of entrypoints bug:
+				const pathToReplace = fileURLToPath(new URL(_config.outDir));
+				newEntryDir = fileURLToPath(new URL(_buildConfig.server, _config.outDir));
+
+				entryPoints.forEach((value, key) => {
+					_entryPoints.set(key, new URL(value.href.replace(pathToReplace, newEntryDir)));
+				});
+			},
 			'astro:build:done': async ({ pages, routes, dir }) => {
-				const entryPath = fileURLToPath(new URL(_buildConfig.serverEntry, _buildConfig.server));
 				const entryUrl = new URL(_buildConfig.serverEntry, _config.outDir);
 				const buildPath = fileURLToPath(entryUrl);
 				// A URL for the final build path after renaming
 				const finalBuildUrl = pathToFileURL(buildPath.replace(/\.mjs$/, '.js'));
+				const functionsUrl = new URL('functions/', _config.root);
 
-				await esbuild.build({
-					target: 'es2020',
-					platform: 'browser',
-					conditions: ['workerd', 'worker', 'browser'],
-					entryPoints: [entryPath],
-					outfile: buildPath,
-					allowOverwrite: true,
-					format: 'esm',
-					bundle: true,
-					minify: _config.vite?.build?.minify !== false,
-					banner: {
-						js: SHIM,
-					},
-					logOverride: {
-						'ignored-bare-import': 'silent',
-					},
-				});
+				if (isModeDirectory) {
+					await fs.promises.mkdir(functionsUrl, { recursive: true });
+				}
+
+				if (isModeDirectory && _buildConfig.split) {
+					const entryPaths = [..._entryPoints.values()].map((entry) => fileURLToPath(entry));
+					const outputDir = fileURLToPath(new URL('.astro', _buildConfig.server));
+					await esbuild.build({
+						target: 'es2020',
+						platform: 'browser',
+						conditions: ['workerd', 'worker', 'browser'],
+						entryPoints: entryPaths,
+						outdir: outputDir,
+						allowOverwrite: true,
+						format: 'esm',
+						bundle: true,
+						minify: _config.vite?.build?.minify !== false,
+						banner: {
+							js: SHIM,
+						},
+						logOverride: {
+							'ignored-bare-import': 'silent',
+						},
+					});
+
+					for (const [key, value] of _entryPoints.entries()) {
+						const newFileName = key.component
+							.replace('src/pages/', '')
+							.replace('.astro', '.js');
+						const newFileUrl = new URL(newFileName, functionsUrl);
+						const newFileDir = path.dirname(fileURLToPath(newFileUrl));
+						const oldFileUrl = new URL(value.href.replace(`${newEntryDir}src/pages`, outputDir).replace('.mjs', '.js'))
+						if (!fs.existsSync(newFileDir)) {
+							fs.mkdirSync(newFileDir, { recursive: true });
+						}
 
-				// Rename to worker.js
-				await fs.promises.rename(buildPath, finalBuildUrl);
+						await fs.promises.rename(oldFileUrl, newFileUrl);
+					}
+				} else {
```

Comment:
this is basically the only new logic here, to support experimental (though not 100%) support.